src/frontends/onnx/frontend/src/op/com.microsoft/matmulnbits.cpp

```diff
@@ -0,0 +1,228 @@
+        CHECK_VALID_NODE(node,
+                         group_idx.get_element_type() == ov::element::i32,
+                         "Unsupported input group_idx type, accepted I32, got: ",
+                         group_idx.get_element_type());
+    }
+
+    if (inputs.size() > 5) {
+        bias = inputs[5];
+        CHECK_VALID_NODE(node,
+                         bias.get_element_type() == a.get_element_type(),
+                         "Unsupported input bias type, must be equal to input A type, got: ",
+                         bias.get_element_type());
+        CHECK_VALID_NODE(node,
+                         bias.get_partial_shape() == PartialShape{N},
+                         "Wrong bias shape, expected [",
+                         N,
+                         "], got: ",
+                         bias.get_partial_shape());
+    }
+
+    {
+        const auto b_const = std::dynamic_pointer_cast<v0::Constant>(b_quantized.get_node_shared_ptr());
+
+        ov::Output<ov::Node> casted_b;
+        ov::Shape casted_b_shape;
+        ov::Output<ov::Node> default_zp;
+        // Casting/converting data of source constant.
+        // For further calculations (sub and/or multiply) we need to reshape it from [N][n_blocks_per_col][blob_size *
+        // X] to [N * n_blocks_per_col][blob_size * X] (where X is amount of values in 1 byte) because scale and
+        // zero_point are represented as: ...with shape like: [N * n_blocks_per_col]...
+        switch (bits) {
+        case 2:
+            casted_b_shape = ov::Shape{static_cast<size_t>(N * n_blocks_per_col), static_cast<size_t>(blob_size * 4)};
+            casted_b = std::make_shared<v0::Constant>(ov::element::u2, casted_b_shape, b_const->get_data_ptr());
+            default_zp = std::make_shared<v0::Constant>(a.get_element_type(), Shape{}, 2);
+            break;
+        case 4:
+            casted_b_shape = ov::Shape{static_cast<size_t>(N * n_blocks_per_col), static_cast<size_t>(blob_size * 2)};
+            casted_b = std::make_shared<v0::Constant>(ov::element::u4, casted_b_shape, b_const->get_data_ptr());
+            default_zp = std::make_shared<v0::Constant>(a.get_element_type(), Shape{}, 8);
+            break;
+        case 8:
+            casted_b_shape = ov::Shape{static_cast<size_t>(N * n_blocks_per_col), static_cast<size_t>(blob_size)};
+            casted_b = op::util::reshape(b_const, casted_b_shape);
+            default_zp = std::make_shared<v0::Constant>(a.get_element_type(), Shape{}, 128);
+            break;
+        default:
+            FRONT_END_THROW("Unsupported bits count");
+            break;
+        }
+
+        // Possible issue with slice implementation, had to move convertion before slice, instead of slicing uint4
+        const auto converted_b = std::make_shared<v1::ConvertLike>(casted_b, a);
+
+        // Simple case
+        if (n_blocks_per_col == 1) {
+            // Removing unused items in case block is bigger than column count
```

Comment:
Could you put more description in the comments to help follow the blocking schema and what we are trying to achieve with the decomposition?

Reply:
B contains a quantized data blocks. Block size should be power of 2 and not less than 16. In case it isn't aligned - rest of a block is filled by zeros. For example, if data is (uint8)[1,2,3,4,5,6] then block will be (uint8)[1,2,3,4,5,6,0,0,0,0,0,0,0,0,0,0]. And last zeros are unused.